oeps/oep-0012.rst

```diff
@@ -0,0 +1,305 @@
+=================================
+OEP-12: Pluggable User Interfaces
+=================================
+
+.. This is the template to use when you start a new OEP.
+
++---------------+-------------------------------------------+
+| OEP           | :doc:`OEP-12 </oeps/oep-0012>`            |
++---------------+-------------------------------------------+
+| Title         | Pluggable User Interfaces                 |
++---------------+-------------------------------------------+
+| Last Modified | 2016-12-06                                |
++---------------+-------------------------------------------+
+| Author        | Andy Armstrong <andya@edx.org>            |
++---------------+-------------------------------------------+
+| Arbiter       | <OEP czar's real name>                    |
++---------------+-------------------------------------------+
+| Status        | Draft                                     |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2016-12-06                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+There is a long standing desire to allow developers to extend edx-platform by
+providing plugins, rather than by modifying the platform's code directly.
+XBlocks provide this ability for new types of courseware components, but
+courseware is only one of many ways in which developers need to extend the
+platform (see `Feature Plugins for edX Platform`_ for a detailed enumeration).
+UI Blocks are a proposed mechanism to allow developers to create pluggable user
+interfaces in Python and install them into edx-platform or other such code
+bases.
```

Comment:
What "other such code bases" do you have in mind? Open edX IDAs? Any python learning-related web apps? Any web/mobile app of any kind? I think that we should have a very clear answer to that question.

My understanding is that XBlocks were designed so that in theory they [could be used on many different platforms to create many types of online educational experience](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/overview/introduction.html#overview) but in practice because they depend on Python and implementing an XBlock runtime is a bit obscure, the edX LMS/CMS is the only place where they are used. (Excluding xblock-sdk which is not an end-user platform.)

Since the XBlock API is not specific to the edX LMS/CMS and the set of available "services" is also limited, there are many things that one wants to do when creating courseware components that XBlocks cannot do. So today we have many XBlocks that don't conform to the XBlock spec because they are just calling into the LMS/CMS internal API directly (and [other issues](https://openedx.atlassian.net/wiki/display/~dormsbee/XBlock+Runtime+Isolation)). I could imagine the same thing happening with UI blocks.